frame/nis/src/lib.rs

```diff
@@ -0,0 +1,937 @@
+						});
+						(0, amount - bid.amount)
+					} else {
+						q.try_insert(0, bid).expect("verified queue was not full above. qed.");
+						(1, amount)
+					};
+
+					let sorted_item_count = q.len().saturating_sub(T::FifoQueueLen::get() as usize);
+					if sorted_item_count > 1 {
+						q[0..sorted_item_count].sort_by_key(|x| x.amount);
+					}
+
+					Ok(net)
+				},
+			)?;
+			QueueTotals::<T>::mutate(|qs| {
+				qs.bounded_resize(queue_count, (0, Zero::zero()));
+				qs[queue_index].0 += net.0;
+				qs[queue_index].1.saturating_accrue(net.1);
+			});
+			Self::deposit_event(Event::BidPlaced { who, amount, duration });
+
+			Ok(())
+		}
+
+		/// Retract a previously placed bid.
+		///
+		/// Origin must be Signed, and the account should have previously issued a still-active bid
+		/// of `amount` for `duration`.
+		///
+		/// - `amount`: The amount of the previous bid.
+		/// - `duration`: The duration of the previous bid.
+		#[pallet::weight(T::WeightInfo::retract_bid(T::MaxQueueLen::get()))]
+		pub fn retract_bid(
+			origin: OriginFor<T>,
+			#[pallet::compact] amount: BalanceOf<T>,
+			duration: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let queue_count = T::QueueCount::get() as usize;
+			let queue_index = duration.checked_sub(1).ok_or(Error::<T>::DurationTooSmall)? as usize;
+			ensure!(queue_index < queue_count, Error::<T>::DurationTooBig);
+
+			let bid = Bid { amount, who };
+			let new_len = Queues::<T>::try_mutate(duration, |q| -> Result<u32, DispatchError> {
+				let pos = q.iter().position(|i| i == &bid).ok_or(Error::<T>::NotFound)?;
+				q.remove(pos);
+				Ok(q.len() as u32)
+			})?;
+
+			QueueTotals::<T>::mutate(|qs| {
+				qs.bounded_resize(queue_count, (0, Zero::zero()));
+				qs[queue_index].0 = new_len;
+				qs[queue_index].1.saturating_reduce(bid.amount);
+			});
+
+			T::Currency::unreserve(&bid.who, bid.amount);
+			Self::deposit_event(Event::BidRetracted { who: bid.who, amount: bid.amount, duration });
+
+			Ok(())
+		}
+
+		/// Ensure we have sufficient funding for all potential payouts.
+		///
+		/// - `origin`: Must be accepted by `FundOrigin`.
+		#[pallet::weight(T::WeightInfo::fund_deficit())]
+		pub fn fund_deficit(origin: OriginFor<T>) -> DispatchResult {
+			T::FundOrigin::ensure_origin(origin)?;
+			let summary: SummaryRecordOf<T> = Summary::<T>::get();
+			let our_account = Self::account_id();
+			let issuance = Self::issuance_with(&our_account, &summary);
+			let deficit = issuance.required.saturating_sub(issuance.holdings);
+			ensure!(!deficit.is_zero(), Error::<T>::Funded);
+			T::Deficit::on_unbalanced(T::Currency::deposit_creating(&our_account, deficit));
+			Self::deposit_event(Event::<T>::Funded { deficit });
+			Ok(())
+		}
+
+		/// Reduce or remove an outstanding receipt, placing the according proportion of funds into
+		/// the account of the owner.
+		///
+		/// - `origin`: Must be Signed and the account must be the owner of the receipt `index` as
```

Comment:
If eventually you need both the receipt and the counterpart, what use cases can the transferrability of the counterpart have?

Reply:
I would not like to speculate.